File: Empresa.java
package libClases;

import java.util.Scanner;

public class Empresa implements Cloneable, Proceso {
    
    // --- ATRIBUTOS ---
    // Usamos un array nativo y lo redimensionamos manualmente
    private Cliente[] clientes;
    private int nClientes; // Número real de clientes guardados
    private int NMax;      // Capacidad actual del array
    private final int INCREMENTO = 5; // Cuánto crece el array cada vez

    // --- CONSTRUCTORES ---
    public Empresa() {
        this.clientes = new Cliente[INCREMENTO];
        this.nClientes = 0;
        this.NMax = INCREMENTO;
    }

    // --- CLONE (Gestión de memoria dinámica manual) ---
   
    public Object clone() {
        Empresa obj = null;
        try {
            obj = (Empresa) super.clone();
            // 1. Clonamos el array (el contenedor)
            obj.clientes = this.clientes.clone();
            
            // 2. Clonamos CADA cliente de dentro (Deep Copy)
            // Si no hacemos esto, la empresa clonada apuntaría a los mismos clientes (punteros compartidos)
            for (int i = 0; i < nClientes; i++) {
                obj.clientes[i] = (Cliente) this.clientes[i].clone();
            }
        } catch (CloneNotSupportedException ex) {
            System.out.println("No se ha podido clonar \n");
        }
        return obj;
    }

    // --- MÉTODOS DE GESTIÓN (ALTA/BAJA) ---

    // Método auxiliar para buscar la posición de un cliente
    public final int buscarCliente(String dni) {
        for (int i = 0; i < nClientes; i++) {
            if (clientes[i].getNif().equals(dni)) {
                return i; // Lo hemos encontrado en la posición i
            }
        }
        return -1; // No existe
    }

    // ALTA: Añadir cliente (con redimensionado dinámico tipo realloc)
    public void alta(Cliente c) {
        int pos = buscarCliente(c.getNif());
        
        if (pos == -1) { // Solo si no existe ya
            // Si el array está lleno, lo ampliamos
            if (nClientes == NMax) {
                NMax += INCREMENTO;
                Cliente[] tablaNueva = new Cliente[NMax];
                
                // Copiamos los punteros del array viejo al nuevo
                for (int i = 0; i < nClientes; i++) {
                    tablaNueva[i] = clientes[i];
                }
                clientes = tablaNueva; // Cambiamos la referencia al nuevo array más grande
            }
            
            // Añadimos el nuevo cliente al final
            clientes[nClientes] = c;
            nClientes++;
        } 
        
    }

    // BAJA: Eliminar cliente (moviendo los huecos a la izquierda)
    public void baja(String dni) {
        int pos = buscarCliente(dni);
        if (pos != -1) {
            // Desplazamos todos los clientes una posición a la izquierda para tapar el hueco
            for (int i = pos; i < nClientes - 1; i++) {
                clientes[i] = clientes[i + 1];
            }
            // Borramos la última referencia duplicada y bajamos el contador
            nClientes--;
            clientes[nClientes] = null; // Limpieza (buena práctica en Java)
        } 
        
    }

    // --- MÉTODOS INTERACTIVOS (PIDEN DATOS AL USUARIO) ---

    public void alta() {
        Scanner s = new Scanner(System.in);
        System.out.print("DNI: ");
        String dni = s.nextLine();
        
        if (buscarCliente(dni) != -1) {
            System.out.println("Ya existe un cliente con este DNI\n" + clientes[buscarCliente(dni)] + "\n");
        } else {
            System.out.print("Indique el nombre del cliente: ");
            String nombre = s.nextLine();
            
            System.out.println("Fecha de nacimiento: "); // Fecha.pedirFecha ya tiene sus propios prints
            Fecha fNac = Fecha.pedirFecha();
            
            System.out.println("Fecha de alta: ");
            Fecha fAlta = Fecha.pedirFecha();
            
            System.out.print("Indique los minutos que habla al mes: ");
            float min = s.nextFloat(); // OJO: Usar coma decimal en consola española
            
            System.out.print("Indique el tipo de cliente (1-Móvil, 2-Tarifa Plana): ");
            int tipo = s.nextInt();
            
            // LIMPIEZA DE BUFFER: Después de leer números (nextInt/Float), queda un Enter colgado.
            // Si viene un nextLine luego, se lo come. Limpiamos aquí por si acaso.
            s.nextLine(); 

            if (tipo == 1) {
                System.out.print("Indique el precio por minuto: ");
                float precio = s.nextFloat();
                
                System.out.println("Fecha fin permanencia: ");
                Fecha fPerman = Fecha.pedirFecha();
                
                // Creamos el objeto y lo mandamos al otro método alta
                alta(new ClienteMovil(dni, nombre, fNac, fAlta, fPerman, min, precio));
                
            } else if (tipo == 2) {
            	
                System.out.print("Nacionalidad: ");
                String nacionalidad = s.nextLine(); // Aquí es donde fallaría si no hubiéramos limpiado el buffer antes
                
                alta(new ClienteTarifaPlana(dni, nombre, fNac, fAlta, min, nacionalidad));
            }
        }
        //s.close(); //No cerramos el Scanner 's' para no cerrar System.in
    }

    public void baja() {
        Scanner s = new Scanner(System.in);
        // CAMBIO 1: Texto exacto del PDF 
        System.out.print("Introduzca nif cliente a dar de baja: ");
        String dni = s.nextLine();
        
        int pos = buscarCliente(dni);
        if (pos != -1) {
            System.out.println(clientes[pos]); // Muestra el cliente encontrado
            
            char confirm;
            do {
                // CAMBIO 2: Texto exacto con interrogación al principio [cite: 78]
                System.out.print("¿Seguro que desea eliminarlo (s/n)? ");
                // next().charAt(0) lee el carácter, pero deja un salto de línea. 
                // No pasa nada porque el bucle limpia o el método acaba.
                confirm = s.next().charAt(0); 
            } while (confirm != 's' && confirm != 'S' && confirm != 'n' && confirm != 'N');
            
            if (confirm == 's' || confirm == 'S') {
                // TRUCO IMPORTANTE: Guardamos el nombre ANTES de borrar.
                // Si borras primero con baja(dni), el array se mueve y clientes[pos] 
                // pasaría a ser el siguiente cliente o null.
                String nombreBorrado = clientes[pos].getNombre();
                
                baja(dni); // Borramos lógicamente
                
                // CAMBIO 3: Mensaje detallado como pide el PDF 
                System.out.println("El cliente " + nombreBorrado + " con nif " + dni + " ha sido eliminado");
            } else {
                // CAMBIO 4: Mensaje con NIF como pide el PDF 
                System.out.println("El cliente con nif " + dni + " no se elimina");
            }
        } else {
            // Este mensaje no sale explícito en la salida del PDF de ejemplo, 
            // pero tu versión es correcta y clara.
            System.out.println("Este DNI no existe \n");
        }
        // No cerramos el scanner s para no matar System.in
    }
    // --- CÁLCULOS Y CONSULTAS ---

    public void descuento(int dto) {
        for (int i = 0; i < nClientes; i++) {
            // RTTI / Dynamic Cast seguro
            if (clientes[i] instanceof ClienteMovil) {
                ClienteMovil cm = (ClienteMovil) clientes[i];
                // CORRECCIÓN: Usamos getPrecioMinuto/setPrecioMinuto (nombres de nuestra clase)
                float nuevoPrecio = cm.getPrecioMinuto() * (1 - dto / 100f);
                cm.setPrecioMinuto(nuevoPrecio);
            }
        }
    }

    public float factura() {
        float total = 0;
        for (int i = 0; i < nClientes; i++) {
            total += clientes[i].factura(); // Polimorfismo
        }
        return total;
    }
    
    public int getN() {
        return nClientes;
    }

    public int nClienteMovil() {
        int n = 0;
        for (int i = 0; i < nClientes; i++) {
            if (clientes[i] instanceof ClienteMovil) {
                n++;
            }
        }
        return n;
    }

    // --- VISUALIZACIÓN ---

   
    public void ver() {
        System.out.println(this.toString());
    }

    
    public String toString() {
        // StringBuilder es mucho más eficiente que concatenar Strings con +
        StringBuilder s = new StringBuilder();
        for (int i = 0; i < nClientes; i++) {
            s.append(clientes[i].toString()).append("\n");
        }
        return s.toString();
    }
}

File: Empresa.cpp
#include "Empresa.h"
#include <iostream>
#include <iomanip>  // Para setprecision
#include <typeinfo> // Para el Paso 4 (typeid)


using namespace std;





// CONSTRUCTOR
Empresa::Empresa() {
    this->nClientes = 0;
    this->nContratos = 0;

    // Inicializa el array dinámico con tamaño 10
    this->capContratos = 10;
    this->contratos = new Contrato*[this->capContratos];
}

//DESTRUCTOR
Empresa::~Empresa() {
    // 1. Libera la memoria de CADA objeto Contrato
    for (int i = 0; i < this->nContratos; i++) {
        delete this->contratos[i]; // Borra el ContratoTP o ContratoMovil
        this->contratos[i] = NULL; // Buena práctica
    }

    // 2. Libera la memoria del array de punteros
    delete [] this->contratos; // El array 'clientes[100]' se destruye solo (memoria estática)
}


Empresa::Empresa(const Empresa& otra)
{
    for (int i=0; i<otra.nClientes; i++)
    {
        this->clientes[i]=otra.clientes[i];
    }
    this->nClientes=otra.nClientes;
    this->nContratos=otra.nContratos;
    this->capContratos=otra.capContratos;

    this->contratos=new Contrato* [this->capContratos];

    for (int i=0; i<this->nContratos; i++)
    {
        ContratoMovil* cm = dynamic_cast<ContratoMovil*>(otra.contratos[i]);

        if (cm !=NULL)
        {
            this->contratos[i]=new ContratoMovil(*cm);
        }
        else
        {
            ContratoTP* cTP = dynamic_cast <ContratoTP*> (otra.contratos[i]);
            if (cTP!=NULL)
            {
                this->contratos[i]=new ContratoTP (*cTP);
            }

        }



    }











}

// Métodos privados

void Empresa::redimensionar() {
    // 1. Duplica la capacidad
    int nuevaCapacidad = this->capContratos * 2;

    // 2. Crea un nuevo array con el doble de capacidad
    Contrato** nuevoArr = new Contrato*[nuevaCapacidad];

    // 3. Copia los punteros antiguos al array nuevo
    for (int i = 0; i < this->nContratos; i++) {
        nuevoArr[i] = this->contratos[i];
    }

    // 4. Borra el array antiguo (SOLO el array, no los 'delete' de los objetos contrato)
    delete [] this->contratos;

    // 5. Apunta al nuevo array
    this->contratos = nuevoArr;

    // 6. Actualiza la capacidad
    this->capContratos = nuevaCapacidad;
}

int Empresa::buscarCliente(long int dni) {//Buscar un cliente por dni
    for (int i = 0; i < this->nClientes; i++) {
        // Usamos . (punto) porque clientes[i] es un OBJETO
        if (this->clientes[i].getDni() == dni) {
            return i; // Devuelve la posición si lo encuentra
        }
    }
    return -1; // No se encontró
}

int Empresa::buscarContrato(int id) {//Buscar un contrato por ID en el array de punteros
    for (int i = 0; i < this->nContratos; i++) {
        // Usamos -> (flecha) porque contratos[i] es un PUNTERO
        if (this->contratos[i]->getIdContrato() == id) {
            return i; // Devuelve la posición si lo encuentra
        }
    }
    return -1; // No se encontró
}


//MÉTODOS PÚBLICOS


void Empresa::cargarDatos() {
    // --- Clientes de prueba ---
    // (Usamos el operador '=' que nos da C++ por defecto para la clase Cliente,
    // que funciona bien porque Cliente gestiona su memoria con su constructor de copia y operador=)

    clientes[0] = Cliente(75547001, "Peter Lee", Fecha(28, 2, 2001));
    clientes[1] = Cliente(45999000, "Juan Perez", Fecha(29, 2, 2000));
    clientes[2] = Cliente(37000017, "Luis Bono", Fecha(31, 1, 2002));
    this->nClientes = 3;

    // --- Contratos de prueba ---
    // (Asegúrate de que la capacidad 'capContratos' es al menos 10)
    contratos[0] = new ContratoMovil(75547001, Fecha(28, 2, 2001), 0.12, 110, "DANES");
    contratos[1] = new ContratoMovil(75547001, Fecha(31, 1, 2002), 0.09, 170, "DANES");
    contratos[2] = new ContratoTP(37000017, Fecha(1, 2, 2002), 250);
    contratos[3] = new ContratoTP(75547001, Fecha(28, 2, 2001), 312);
    contratos[4] = new ContratoMovil(45999000, Fecha(31, 1, 2002), 0.10, 202, "ESPAÑOL");
    contratos[5] = new ContratoMovil(75547001, Fecha(31, 1, 2002), 0.15, 80, "DANES");
    contratos[6] = new ContratoTP(45999000, Fecha(1, 2, 2002), 400);
    this->nContratos = 7;
}

void Empresa::ver() const {
    cout << "--- Clientes: " << this->nClientes << endl;
    cout << "--- Contratos: " << this->nContratos << endl;
    // 1. Imprimir "La Empresa tiene X clientes y Y contratos"
    cout << "La Empresa tiene " << this->nClientes << " clientes y " << this->nContratos << " contratos" << endl;

    // 2. Imprimir "Clientes:" y un bucle 'for' que llama a 'cout << clientes[i]'
    cout << "Clientes:" << endl;
    for (int i = 0; i < this->nClientes; i++) {
        cout << this->clientes[i] << endl;
    }

    // 3. Imprimir "Contratos:" y un bucle 'for' que llama a 'cout << *contratos[i]'
    cout << "Contratos:" << endl;
    for (int i = 0; i < this->nContratos; i++) {
        // Llama al operator<<(Contrato) -> que llama a c.imprimir() (VIRTUAL)
        cout << *(this->contratos[i]) << endl;
    }
}

void Empresa::crearContrato() {
    long int dni;
    int dia, mes, anio;

    cout << "Introduzca dni: ";
    cin >> dni;

    int posCliente = this->buscarCliente(dni);

    // 3. Si no existe (-1)
    if (posCliente == -1) {
        // a. Comprobar si 'nClientes < 100'
        if (this->nClientes >= 100) {
            cout << "Error: Límite de clientes alcanzado." << endl;
            return;
        }

        // b. Pedir Nombre y Fecha
        char nombre[100];
        cout << "Nombre del cliente: ";
        cin.ignore(); // Limpia el buffer del 'cin' anterior
        cin.getline(nombre, 100);

        cout << "Fecha de alta (dd mm aaaa): ";
        cin >> dia >> mes >> anio;

        // c. Crear Cliente nuevo en 'clientes[nClientes]' y hacer nClientes++
        this->clientes[this->nClientes] = Cliente(dni, nombre, Fecha(dia, mes, anio));
        posCliente = this->nClientes; // La posición del nuevo cliente es la última
        this->nClientes++;

    }

    // 4. Si existe (o se acaba de crear)

    // a. Comprobar si 'nContratos == capContratos' y llamar a redimensionar() (si el array dinámico está lleno)
    if (this->nContratos == this->capContratos) {
        this->redimensionar();
    }

    // b. Pedir tipo (1-TP, 2-Movil)
    int tipo;
    cout << "Tipo de Contrato a abrir (1-Tarifa Plana, 2-Movil): ";
    cin >> tipo;

    // c. Pedir datos comunes (fecha)
    cout << "Fecha del contrato" << endl;
    cout << "dia: "; cin >> dia;
    cout << "mes: "; cin >> mes;
    cout << "anio: "; cin >> anio;
    Fecha fContrato(dia, mes, anio);

    // Crea el objeto Contrato (TP o Movil) con 'new''
    if (tipo == 1) {
        int min;
        cout << "minutos hablados: ";
        cin >> min;
        this->contratos[this->nContratos] = new ContratoTP(dni, fContrato, min);

    } else if (tipo == 2) {
        int min;
        float precio;
        char nac[100];
        cout << "minutos hablados: ";
        cin >> min;
        cout << "Precio minuto: ";
        cin >> precio;
        cout << "Nacionalidad: ";
        cin >> nac;
        this->contratos[this->nContratos] = new ContratoMovil(dni, fContrato, precio, min, nac);
    }

    // e. nContratos++
    this->nContratos++;
}

bool Empresa::cancelarContrato(int idContrato) {
    // 1. Llamar a buscarContrato(idContrato)
    int pos = this->buscarContrato(idContrato);

    // 2. Si no existe (-1)
    if (pos == -1) {
        // cout << "El Contrato ... no existe" (lo hace el main)
        return false;
    }

    // 3. Si existe (en la pos 'pos'):
    // a. 'delete contratos[pos];' (¡Libera la memoria!)
    delete this->contratos[pos];

    // b. Mover todos los punteros desde 'pos+1' para tapar el hueco
    for (int i = pos; i < this->nContratos - 1; i++) {
        this->contratos[i] = this->contratos[i + 1];
    }

    // c. 'nContratos--'
    this->nContratos--;

    // d. 'return true'
    return true;
}


bool Empresa::bajaCliente(long int dni) {
    // 1. Llamar a buscarCliente(dni)
    int posCliente = this->buscarCliente(dni);

    // 2. Si no existe (-1)
    if (posCliente == -1) {
        // cout << "El cliente ... no existe" (lo hace el main)
        return false;
    }

    // 3. Si existe (en la pos 'posCliente'):

    // a. Recorrer 'contratos' y eliminar los suyos
    // b. ¡OJO! Al borrar, el array se encoge. Hay que reajustar el 'for' (i--)
    for (int i = 0; i < this->nContratos; i++) {
        if (this->contratos[i]->getDniContrato() == dni) {
            cancelarContrato(this->contratos[i]->getIdContrato());
            i--; // Decrementa 'i' para volver a comprobar la posición actual
        }
    }
    /*
     Borrar de un array mientras se itera
      Recorremos el array y, si borramos un contrato,
      llamamos a cancelarContrato().
      cancelarContrato() encoge el array, así que tenemos
      que decrementar 'i' (i--) para volver a comprobar
      la posición actual (que ahora contiene un nuevo contrato).
     */

    // c. Borra al cliente del array 'clientes' (tapando el hueco)
    //    (Recuerda: aquí asignamos objetos, no punteros)
    for (int i = posCliente; i < this->nClientes - 1; i++) {
        this->clientes[i] = this->clientes[i + 1];
    }

    // d. 'nClientes--'
    this->nClientes--;

    // e. 'return true'
    return true;
}

/*'typeid'
  Comprobación de tipo EXACTA.
  Se usa para contar.
  Necesita <typeinfo>
  Se usa typeid(*puntero), con el asterisco.
 */
int Empresa::nContratosTP() const {
    int contador = 0;

    // 2. Bucle 'for' de 0 a 'nContratos'
    for (int i = 0; i < this->nContratos; i++) {
        // 3. 'if (typeid(*contratos[i]) == typeid(ContratoTP)) { contador++; }'
        if (typeid(*(this->contratos[i])) == typeid(ContratoTP)) {
            contador++;
        }
    }

    // 4. 'return contador;'
    return contador;
}

void Empresa::descuento(float porcentaje) {//Función implementada para el apartado 4

    // 1. Bucle 'for' de 0 a 'nContratos'
    for (int i = 0; i < this->nContratos; i++) {

        // 2. 'ContratoMovil* cm = dynamic_cast<ContratoMovil*>(contratos[i]);'
        // Intenta convertir Contrato* -> ContratoMovil*
        ContratoMovil* cm = dynamic_cast<ContratoMovil*>(this->contratos[i]);

        // 3. 'if (cm != NULL) { ... }' // Comprueba si la conversión tuvo éxito
        if (cm != NULL) {
            // a. 'float precioActual = cm->getPrecioMinuto();'
            float precioActual = cm->getPrecioMinuto();

            // b. 'cm->setPrecioMinuto(...)'
            cm->setPrecioMinuto(precioActual * (1.0 - (porcentaje / 100.0)));
        }
    }
}


void Empresa::mejorClienteContratoTP() const
{
    long int dni_maxtp=-1;
    int n_tp_max=0;


    for (int i=0; i<nClientes; i++)
    {

        long int dni_actual= clientes[i].getDni();

        int n_tp_actual= 0;
        for (int j=0; j<nContratos; j++)
        {

            if (dni_actual==contratos[j]->getDniContrato())
            {
                ContratoTP* ctp = dynamic_cast<ContratoTP*> (contratos[j]);
                if (ctp != NULL)
                    {
                        n_tp_actual++;

                    }
            }

        }
        if (n_tp_actual>=n_tp_max)
        {
            dni_maxtp=dni_actual;
            n_tp_max=n_tp_actual;
        }







    }

    cout<<"Dni del ganador: "<<dni_maxtp<<endl<<"Numero de contratos: "<<n_tp_max<<endl;


}




void Empresa::mejorClienteDeLaEmpresaMejorTotalDeFacturas()
{
    long int dni_max=-1;
    float max_total=0;


    for (int i=0; i<nClientes; i++)
    {
        float total_actual=0;

        long int dni_actual= clientes[i].getDni();

        for (int j=0; j<nContratos; j++)
        {

            if (dni_actual==contratos[j]->getDniContrato())
            {
                ContratoTP* ctp = dynamic_cast<ContratoTP*> (contratos[j]);
                if (ctp != NULL)
                    {
                        total_actual=total_actual+ctp->factura();


                    }
                else
                    {
                        ContratoMovil* cm =dynamic_cast <ContratoMovil*> (contratos[j]);
                        if (cm !=NULL)
                        {
                            total_actual=total_actual+cm->factura();

                        }
                    }

            }

        }
        if (max_total<=total_actual)
        {
            dni_max=dni_actual;
            max_total=total_actual;
        }








    }
    cout<<"Dni del ganador: "<<dni_max<<endl<<"Total gastado: "<<max_total<<endl;


}
